### ISSUE TITLE
Channel 1 not available

### ISSUE THREAD

**Issue by mansuen** [sibartel/lecroyutils#4]
Hi! Foremost, thank you for your work. 

I have two issues when running the example code. First, when I try to run it, it returns me the error:

```
ModuleNotFoundError: No module named 'data'
```

At line:

`from lecroyutils.data import LecroyScopeData
`

I just moved all the code from data to control and the error didn't appear any more. Then, an error appeared at lines:

`data = scope.waveform('C1')`

or

`scope.save_waveform('C1', 'C1_00000_Lecroy.trc')`

And the error was:

`Exception: Channel C1 not available.`

I have my HDO4024A LeCroy oscilloscope receiving a signal on Channel 1, and I tried leaving it on Normal and Stop trigger, but the error would still appear. I tried changing channel but no luck. Do you know what could be happening? Sorry, my knowledge in Python and remote control oscilloscopes is very limited. 

My goal is to simply take a single measurement and save the waveform. The next step would be to take the waveform from a Math channel that takes averages of the measurement of a channel.

Thank you for your help in advance!

**Comment by sibartel** [comment]
Hi mansuen,

regarding your first problem:
- Did you install lecroyutils using pip?
- Which python version do you use?

Currently I don't know why C1 is not available. Maybe your oscilloscope behaves a little different from the one I had at hand. Can you please provide me the output of:
```python
from lecroyutils.control import LecroyScope

scope = LecroyScope('127.0.0.1') # Change this according to your setup

print(scope._read('app.ExecsNameAll'))
```

Looking forward to here from you.
Best regards!

**Comment by mansuen** [comment]
Hi sibartel!

Thank you for your kind reply. Regarding the first part, yes, I installed lecroyutils with pip, and I use Python 3.7.4 with Anaconda 4.8.0. I'm probably making some noob mistake somewhere, but I can't figure where.

The output of that code is:

`2,C3,C4,F1,F2,Z1,Z2,Z3,Z4,M1,M2,M3,M4,P1,P2,P3,P4,P5,P6,P7,P8,Q1,Measure,PassFail,ScanDecode`

It seems that it can't read C1 and C2? I've tried running the code on C3 and it successfully reads and saves the waveform! But I get this error when parsing the waveform:

`UserWarning: Signal was clipped: C1_00001_Lecroy.trc`

What will the parsing exactly do? Will it parse `C1_00001_Lecroy.trc` into a legible data file?

Thanks so much for your help. I would be completely clueless without your program and your help!

**Comment by sibartel** [comment]
Hi mansuen,

perfect, installing lecroyutils using pip is the preferred way. You said that you copied code (from data.py to contro.pyl). Did you do that in the `site-packages` directory of your python installation? Or did you download/clone the lecroyutils src to a new directory and edited the source files there?

Installing lecroyutils via pip is the only thing you have to do. No additional download or local copy of lecroyutils should be necessary.

> It seems that it can't read C1 and C2?

Yes it seems like C1 and C2 are not correctly detected by lecroyutils as valid channels. I suspect that your oscilloscope, unlike the devices I tested, returns the response data directly without a prefix. The expected prefix is 4 characters long and is normally discarded. In your case, this causes the substring `'C1,C` of the response to be mistakenly dropped.

Currently I am unsure which oscilloscope is behaving 'correctly'. Did you update the firmware of your oscilloscope the latest available?

I'll publish an update to lecroyparser in the next days to handle both variants.

> It seems that it can't read C1 and C2? I've tried running the code on C3 and it successfully reads and saves the waveform! But I get this error when parsing the waveform:

That is not an error, just an (experimental!) warning to inform you that the measured signal could be limited/clipped therefore cut-off because of the input measuring range of the oscilloscope. See https://en.wikipedia.org/wiki/Clipping_(signal_processing)

Maybe adjust the measuring range at your oscilloscope or ignore the warning.

> I parsed the file finally with lecroyparser
> `data = lecroyparser.ScopeData('./C1_00001_Lecroy.trc')`

Using [lecroyparser](https://github.com/bennomeier/leCroyParser) here is totally fine. The in lecroyutils integrated parser is actually based on [lecroyparser](https://github.com/bennomeier/leCroyParser) but to my knowledge offers more features (e.g., support for sequence mode) and many more smaller adjustments.

If you are looking for a more lecroyutils way to do that, you can either directly access the waveform data without saving it to a file:
```python
data = scope.waveform('C3')
```
Or parse a saved file:
```python
data = LecroyScopeData.parse_file('C1_00001_Lecroy.trc')
```
After that you can access all properties via the data object:
```python
data.x
data.y
data.x_unit
data.y_unit
data.y_min
data.y_max
...
```

I hope my answer was helpful.
Looking forward to here from you again.
Best regards!

**Comment by mansuen** [comment]
Hi sibartel! Thank you again for being so kind to help me out! I'm sorry for the late reply.

First, yes, I edited data.py and control.py from `site-packages`. I don't understand why this happens really. 

I've checked and the firmware is updated; it is a new HDO4024A oscilloscope. 

I see, I didn't realize I was indeed clipping the signal! Thank you for the explanation.

Fantastic, thank you for the explanation regarding the lecroyutils integrated parser. It works!

Best regards!

**Comment by sibartel** [comment]
Heyho mansuen,

I finally found time this evening to take a closer look at the problems you mentioned and committed two small bug fixes. Could you install the current git version using pip and test the changes? In case everything works I will release a new version with the changes.
```bash
> pip install git+https://github.com/sibartel/lecroyutils.git
```

In case your problems are solved, you could do me a favor by testing the other functions of lecroyutils on your osci:
```bash
from lecroyutils.control import LecroyScope, TriggerMode, TriggerType

scope = LecroyScope('127.0.0.1') # Change this according to your setup
scope.trigger_type = TriggerType.edge
scope.trigger_source = 'C1'
scope.acquire(force=True)
scope.save_screenshot('test.png')
```

Looking forward to here from you again.
Best regards!

**Comment by mansuen** [comment]
Hi sibartel, thank you so much. I'll get back to you after new year when I get access to the oscilloscope again. Happy holidays!

**Comment by sibartel** [comment]
Closed because of inactivity. Changes were merged into [v4.0.0](https://github.com/sibartel/lecroyutils/releases/tag/v4.0.0).
